### PR TITLE
[Docs] Autotuner update

### LINKED ISSUE #1391: Unable to run AutoTuner for gcd
### Subject

[Flow] for any util, flow Makefile, or flow script issues.

### Describe the bug

```
python3 distributed.py --design gcd --platform sky130hd \
                       --config ../designs/sky130hd/gcd/autotuner.json \
                       tune
```
Failing to complete autotuner for gcd design as per document: https://openroad-flow-scripts.readthedocs.io/en/latest/user/InstructionsForAutoTuner.html#how-to-use

### Expected Behavior

Complete AutoTuner successfully.

### Environment

```shell
Git commit: 318f3042c43d9ac0c4eea675fc5f0f709b58fcba
kernel: Linux 3.10.0-1160.90.1.el7.x86_64
os: CentOS Linux 7 (Core)
cmake version 3.24.2
-- The CXX compiler identification is GNU 8.3.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/rh/devtoolset-8/root/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- OpenROAD version: v2.0-9980-g318f3042c
-- System name: Linux
-- Compiler: GNU 8.3.1
-- Build type: RELEASE
-- Install prefix: /usr/local
-- C++ Standard: 17
-- C++ Standard Required: ON
-- C++ Extensions: OFF
-- The C compiler identification is GNU 8.3.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/rh/devtoolset-8/root/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found Python: /usr/local/bin/python3.9 (found version "3.9.6") found components: Interpreter 
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Performing Test C_COMPILER_SUPPORTS__-Wall
-- Performing Test C_COMPILER_SUPPORTS__-Wall - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wall
-- Performing Test CXX_COMPILER_SUPPORTS__-Wall - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-array-bounds
-- Performing Test C_COMPILER_SUPPORTS__-Wno-array-bounds - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-array-bounds
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-array-bounds - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-nonnull
-- Performing Test C_COMPILER_SUPPORTS__-Wno-nonnull - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-nonnull
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-nonnull - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-maybe-uninitialized
-- Performing Test C_COMPILER_SUPPORTS__-Wno-maybe-uninitialized - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-maybe-uninitialized
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-maybe-uninitialized - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format-overflow
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format-overflow - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format-overflow
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format-overflow - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-variable
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-variable - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-variable
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-variable - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-function
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-function - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-function
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-function - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-write-strings
-- Performing Test C_COMPILER_SUPPORTS__-Wno-write-strings - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-write-strings
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-write-strings - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-sign-compare
-- Performing Test C_COMPILER_SUPPORTS__-Wno-sign-compare - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-sign-compare
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-sign-compare - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-deprecated
-- Performing Test C_COMPILER_SUPPORTS__-Wno-deprecated - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-deprecated
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-deprecated - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-c++11-narrowing
-- Performing Test C_COMPILER_SUPPORTS__-Wno-c++11-narrowing - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-c++11-narrowing
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-c++11-narrowing - Failed
-- Performing Test C_COMPILER_SUPPORTS__-Wno-register
-- Performing Test C_COMPILER_SUPPORTS__-Wno-register - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-register
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-register - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal
-- Performing Test C_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal - Failed
-- Performing Test C_COMPILER_SUPPORTS__-fpermissive
-- Performing Test C_COMPILER_SUPPORTS__-fpermissive - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-fpermissive
-- Performing Test CXX_COMPILER_SUPPORTS__-fpermissive - Success
-- Performing Test C_COMPILER_SUPPORTS__-x
-- Performing Test C_COMPILER_SUPPORTS__-x - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-x
-- Performing Test CXX_COMPILER_SUPPORTS__-x - Failed
-- Performing Test C_COMPILER_SUPPORTS__c++
-- Performing Test C_COMPILER_SUPPORTS__c++ - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__c++
-- Performing Test CXX_COMPILER_SUPPORTS__c++ - Failed
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-but-set-variable
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-but-set-variable - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-but-set-variable
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-but-set-variable - Success
-- TCL library: /usr/lib64/libtcl.so
-- TCL header: /usr/include/tcl.h
-- TCL readline library: /usr/lib64/libtclreadline.so
-- TCL readline header: /usr/include
-- Found SWIG: /usr/bin/swig (found suitable version "4.1.0", minimum required is "3.0")  
-- Using SWIG >= 4.1.0 -flatstaticmethod flag for python
-- Found Boost: /usr/local/lib/cmake/Boost-1.72.0/BoostConfig.cmake (found version "1.72.0")  
-- boost: 1.72.0
-- Found Python3: /usr/local/include/python3.9 (found version "3.9.6") found components: Development Development.Module Development.Embed 
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.7") 
-- spdlog: 1.8.1
-- Found BISON: /usr/bin/bison (found version "3.0.4") 
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.5") found components: doxygen dot 
-- STA version: 2.4.0
-- STA git sha: 7cf916ba205115a06c4531a044ced481f1ff8f12
-- System name: Linux
-- Compiler: GNU 8.3.1
-- Build type: RELEASE
-- Build CXX_FLAGS: -O3 -DNDEBUG
-- Install prefix: /usr/local
-- Found FLEX: /usr/local/bin/flex (found version "2.6.4") 
-- TCL library: /usr/lib64/libtcl.so
-- TCL header: /usr/include/tcl.h
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- SSTA: 0
-- STA executable: /home/vijayan/OPENROAD_FLOW/ORFS_LOCAL/OpenROAD-flow-scripts/tools/OpenROAD/src/sta/app/sta
-- GPU is not enabled
-- TCL library: /usr/lib64/libtcl.so
-- TCL header: /usr/include/tcl.h
-- Found re2: /opt/or-tools/lib64/cmake/re2/re2Config.cmake (found version "9.0.0") 
-- Found Clp: /opt/or-tools/lib64/cmake/Clp/ClpConfig.cmake (found version "1.17.7") 
-- Found Cbc: /opt/or-tools/lib64/cmake/Cbc/CbcConfig.cmake (found version "2.10.7") 
-- Found Eigen3: /usr/local/share/eigen3/cmake/Eigen3Config.cmake (found version "3.3.90") 
-- Found SCIP: /opt/or-tools/lib/cmake/scip/scip-config.cmake (found version "8.0.1") 
-- GUI is enabled
-- Charts widget is not enabled
-- Found Boost: /usr/local/lib/cmake/Boost-1.72.0/BoostConfig.cmake (found version "1.72.0") found components: serialization 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Could NOT find VTune (missing: VTune_LIBRARIES VTune_INCLUDE_DIRS) 
-- Found Boost: /usr/local/lib/cmake/Boost-1.80.0/BoostConfig.cmake (found suitable version "1.80.0", minimum required is "1.78")  
-- TCL library: /usr/lib64/libtcl.so
-- TCL header: /usr/include/tcl.h
-- Found Boost: /usr/local/lib/cmake/Boost-1.80.0/BoostConfig.cmake (found version "1.80.0") found components: serialization system thread 
-- TCL readline enabled
-- Tcl Extended disabled
-- Python3 enabled
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/tmp.CEz96EKs67
```

### To Reproduce

```
cd flow/util
python3 distributed.py --design gcd --platform sky130hd \
                       --config ../designs/sky130hd/gcd/autotuner.json \
                       tune
```

### Relevant log output

```shell
Traceback (most recent call last):
  File "/home/vijayan/OPENROAD_FLOW/ORFS_LOCAL/OpenROAD-flow-scripts/flow/util/distributed.py", line 41, in <module>
    import ray
  File "/home/vijayan/.local/lib/python3.9/site-packages/ray/__init__.py", line 91, in <module>
    import ray._raylet  # noqa: E402
  File "python/ray/_raylet.pyx", line 115, in init ray._raylet
  File "/home/vijayan/.local/lib/python3.9/site-packages/ray/exceptions.py", line 7, in <module>
    from ray.core.generated.common_pb2 import RayException, Language, PYTHON
  File "/home/vijayan/.local/lib/python3.9/site-packages/ray/core/generated/common_pb2.py", line 15, in <module>
    from . import runtime_env_common_pb2 as src_dot_ray_dot_protobuf_dot_runtime__env__common__pb2
  File "/home/vijayan/.local/lib/python3.9/site-packages/ray/core/generated/runtime_env_common_pb2.py", line 36, in <module>
    _descriptor.FieldDescriptor(
  File "/home/vijayan/.local/lib/python3.9/site-packages/google/protobuf/descriptor.py", line 561, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates
```

### Screenshots

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by luarss**
Todo
- add images for tensorboard
- @dralabeing to update the text for tune/sweep modes.
fixes #1391

### CODE REVIEW

**Review by dralabeing — APPROVED** [review]

**Comment by luarss** [comment]
@dralabeing Pending your tensorboard image for autotuner metrics

**Comment by luarss** [comment]
@dralabeing There is an error with the documentation compilation in the actions below. It is showing an outdated version. Here is how the installation instructions should look: 

image

**Comment by luarss** [comment]
Added the autotuner images from @vijayank88 

image

**Comment by dralabeing** [comment]
@luarss the images are good to go.

**Comment by luarss** [comment]
@vvbandeira Can you have a look?

File: docs/user/InstructionsForAutoTuner.md [hunk 1]
```diff
@@ -18,14 +18,32 @@ AutoTuner contains top-level Python script for ORFS, each of which implements a
...
+which is a free minimal installer for the package manager `conda`. 
+
+```shell
+# set up conda environment
+conda create -n autotuner_env python=3.7
+conda activate autotuner_env
+
+# install requirements
+cd <PATH_TO_ORFS_ROOT>/tools/AutoTuner
+pip install -r requirements.txt
```
Comment (vvbandeira) [inline docs/user/InstructionsForAutoTuner.md]:
```suggestion
pip install -r ./tools/AutoTuner/requirements.txt
```

File: docs/user/InstructionsForAutoTuner.md [hunk 2]
```diff
@@ -70,8 +88,7 @@ For SDC you can use:
   - I/O delay. This will create a copy of `_SDC_FILE_PATH` and modify the I/O delay.
 
 
-
-For FastRoute you can use:
+For Global Routing (FastRoute) you can use:
```
Comment (vvbandeira) [inline docs/user/InstructionsForAutoTuner.md]:
```suggestion
For Global Routing parameters that are set on `fastroute.tcl` you can use:
```

File: docs/user/InstructionsForAutoTuner.md [hunk 3]
```diff
@@ -83,21 +100,25 @@ For FastRoute you can use:
...
-For both sweep and tune modes <mode>:
-```shell
-python3 distributed.py -h
-```
+In the sweeping mode, we manually specify the parameters search space
+in the `json`; whereas in the tuning mode - we will use Ray's Tune 
+feature to automatically optimise hyperparameters using one of the 
+algorithms listed above with autotuner `json` file. These two files 
+are of very different format, and you may refer to the examples below
+for more information.
```
Comment (vvbandeira) [inline docs/user/InstructionsForAutoTuner.md]:
```suggestion
The `distributed.py` script supports two modes of operation: `sweep`, where every possible parameter combination in the search space is tested, and `tune`, where we use Ray's Tune feature to intelligently search the space and optimise hyperparameters using one of the algorithms listed above. The `sweep` mode is useful when we want to isolate or test a single or very few parameters. On the other hand, `tune` is more suitable for finding the best combination of a complex and large number of flow parameters. Both modes rely on user-specified search space that is defined by a `.json` file, they use the same syntax and format, though some features may not be available for sweeping.
```

Comment (vvbandeira) [reply docs/user/InstructionsForAutoTuner.md]: I think this is a more accurate description of the difference. @dralabeing please check if the wording is clear.

**Review by vvbandeira — CHANGES_REQUESTED** [review]
Just some feedback

File: docs/user/InstructionsForAutoTuner.md (hunk 3 above)
Comment (dralabeing) [reply docs/user/InstructionsForAutoTuner.md]: @vvbandeira . I have a few comments:
1. Both sweep and tune use Ray correct? Also are they both automatic in terms of search? Or is sweep manual based on the full search space defined by the user?
2. What is the difference between the search space by sweep and that of tune? Is one faster or slower? Similar to one but what is the speed impact. 
3. From a  use model perspective , it implies that sweep is faster than tune but may not give accurate results. Is this correct?
4. Why use sweep at all if it does not search intelligently across the design space?
I think we need to clarify usage a bit more along these lines. Also use American english, optimize vs optimise etc.

Comment (vvbandeira) [reply docs/user/InstructionsForAutoTuner.md]: 1a. Yes, both `sweep` and `tune` use Ray either for job distribution (both) or search (only tune).
1b. `tune` is ML search on user-defined search space. `sweep` is a "brute force" for-loop-style that goes over all possible combinations in the same kind of user-defined search space. In all cases, the user manually defines the search space via a config `.json` file.
2. As mentioned above, you can think of `tune` as a smart search and `sweep` as a "brute force" exploration (no search is involved).
3. `tune` and `sweep` serve different purposes; we cannot compare them apples-to-apples. `tune` is used when the user wants to know the answer to "What are the values for each parameter in my search space to get the best results from ORFS". While the `sweep` would normally be used to answer, "What happens to my design as the parameter X value is set to {1, 2, ..., 100}?"
4. `sweep` is useful, especially for people developing code in OpenROAD, to find good defaults for knobs or in cases where it is very important that you squeeze and get the absolute best of ORFS regardless of the amount of time and CPU needed.

File: docs/user/InstructionsForAutoTuner.md [hunk 4]
```diff
@@ -106,12 +105,16 @@ For Global Routing (FastRoute) you can use:
...
-In the sweeping mode, we manually specify the parameters search space
-in the `json`; whereas in the tuning mode - we will use Ray's Tune 
-feature to automatically optimise hyperparameters using one of the 
-algorithms listed above with autotuner `json` file. These two files 
-are of very different format, and you may refer to the examples below
-for more information. 
+The `distributed.py` script supports two modes of operation: `sweep`,
+where every possible parameter combination in the search space is tested,
+and `tune`, where we use Ray's Tune feature to intelligently search the
+space and optimise hyperparameters using one of the algorithms listed
```
Comment (vvbandeira) [inline docs/user/InstructionsForAutoTuner.md]:
```suggestion
space and optimize hyperparameters using one of the algorithms listed 
```

File: docs/user/InstructionsForAutoTuner.md [hunk 5]
```diff
@@ -106,12 +105,16 @@ For Global Routing (FastRoute) you can use:
 
 `distributed.py` scripts handles sweeping and tuning of ORFS parameters.
 
-In the sweeping mode, we manually specify the parameters search space
-in the `json`; whereas in the tuning mode - we will use Ray's Tune 
-feature to automatically optimise hyperparameters using one of the 
-algorithms listed above with autotuner `json` file. These two files 
-are of very different format, and you may refer to the examples below
-for more information. 
+The `distributed.py` script supports two modes of operation: `sweep`,
```
Comment (vvbandeira) [inline docs/user/InstructionsForAutoTuner.md]:
```suggestion
The `distributed.py` script leverages Ray's job scheduling and management to fully utilize available hardware, albeit a single local machine or a cloud environment with thousands of CPUs. The two modes of operation are`sweep`,
```

Comment (vvbandeira) [reply docs/user/InstructionsForAutoTuner.md]: Remove line 106 and include the change above

**Review by vvbandeira — CHANGES_REQUESTED** [review]

File: docs/user/InstructionsForAutoTuner.md (hunk 3 above)
Comment (vvbandeira) [reply docs/user/InstructionsForAutoTuner.md]: @dralabeing I have added some text to explain the role of Ray in both use cases. Should Jack or I draft a complete example about the intended usage of `sweep` and `tune`?

File: docs/user/InstructionsForAutoTuner.md (hunk 5 above)
Comment (dralabeing) [reply docs/user/InstructionsForAutoTuner.md]: Clarify sentence to: The `distributed.py` script uses Ray's job scheduling and management to fully utilize available hardware resources from a single server to a distributed server configuration, on-premise or over the cloud with multiple CPUs.

File: docs/user/InstructionsForAutoTuner.md (hunk 3 above)
Comment (dralabeing) [reply docs/user/InstructionsForAutoTuner.md]: Thanks @vvbandeira In 3 above, there is still a qualitative analysis designers make to determine the impact on the design when parameters are swept. How is this analysis done without applying a cost function?  Is this through manual analysis alone i.e viewing each sweep experiment through timing, area, congestion?  Also, what is Ray doing for the sweep - is it simply scheduling and job management without the search? 

I am not entirely clear about the use model and value of sweep vs tune. Let's discuss offline before we finalize doc changes.

**Comment by luarss** [comment]
@vvbandeira @dralabeing Updated comments from both.

**Comment by luarss** [comment]
@vvbandeira Require your input for the writeup

**Comment by vijayank88** [comment]
@luarss @dralabeing 
Is anything else pending on this?

**Comment by luarss** [comment]
@vijayank88 I believe we need Vitor's writeup for sweeping and tuning modes. This is our current writeup:

image

**Review by vvbandeira — APPROVED** [review]

**Pull request merged**